File: python/212_WordSearchII.py
from unittest import TestCase
# https://leetcode.com/problems/word-search-ii/


class Node:
    count = 0

    def __init__(self):
        Node.count += 1
        self.id = Node.count
        self.end = False
        self.next = {}

    def insert(self, word):
        if not word:
            self.end = True
        else:
            ch = word[0]
            if not self.next.get(ch):
                self.next[ch] = Node()
            self.next[ch].insert(word[1:])


class Trie:
    def __init__(self):
        self.root = Node()

    def insert(self, word):
        self.root.insert(word)


class WordSearchII:
    def findWords(self, board, words):
        """
        :type board: List[List[str]]
        :type words: List[str]
        :rtype: List[str]
        """
        trie = Trie()
        for w in words:
            trie.insert(w)

        result = set()
        visited = set()

        def search(node, i, j, pre):
            if not (0 <= i < len(board) and 0 <= j < len(board[i])) or (i, j) in visited:
                return
            ch = board[i][j]
            node = node.next.get(ch)
            if not node:
                return
            pre += ch
            if node.end:
                result.add(pre)
            visited.add((i, j))
            for (ni, nj) in ((i,j+1), (i+1,j), (i,j-1), (i-1,j)):
                search(node, ni, nj, pre)
            visited.remove((i, j))

        for i in range(len(board)):
            for j in range(len(board[i])):
                search(trie.root, i, j, "")
        return list(result)


class Tests(TestCase):
    def test1(self):
        search = WordSearchII()
        self.assertEqual({"eat", "oath"}, set(
            search.findWords(
                [
                    ['o', 'a', 'a', 'n'],
                    ['e', 't', 'a', 'e'],
                    ['i', 'h', 'k', 'r'],
                    ['i', 'f', 'l', 'v']
                ],
                ["oath", "pea", "eat", "rain"]
            )
        ))

    def test2(self):
        search = WordSearchII()
        self.assertEqual([],
            search.findWords(
                [
                    ["a", "a"]
                ],
                ["aaa"]
            )
        )

    def test3(self):
        search = WordSearchII()
        self.assertEqual(["a"],
                         search.findWords(
                             [
                                 ["a"]
                             ],
                             ["a"]
                         ))

File: kotlin/src/main/kotlin/212.WordSearchII.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/word-search-ii/

class WordSearchII{
  class Trie{
    class Node{
      val next = HashMap<Char, Node>()
      var end = false
    }
    val root = Node()

    fun insert(word: String){
      var node = root
      for(ch in word){
        if(!node.next.containsKey(ch))
          node.next[ch] = Node()
        node = node.next[ch]!!
      }
      node.end = true
    }
  }

  fun findWords(board: Array<CharArray>, words: Array<String>): List<String> {
    val trie = Trie()
    for(w in words)
      trie.insert(w)

    val result = HashSet<String>()
    val visited = HashSet<Pair<Int, Int>>()

    fun search(node: Trie.Node, i: Int, j: Int, pre: String){
      if(!(0<=i && i<board.size && 0<=j && j<board[i].size) || visited.contains(i to j))
        return
      val ch = board[i][j]
      val cur = pre+ch
      val nd =  node.next.get(ch)
      if(nd == null) return
      if(nd.end)
        result.add(cur)

      visited.add(i to j)
      for((ni, nj) in listOf(i to j+1, i+1 to j, i to j-1, i-1 to j))
        search(nd, ni, nj, cur)
      visited.remove(i to j)
    }

    for(i in 0 until board.size)
      for(j in 0 until board[i].size)
        search(trie.root, i, j, "")

    return result.toList()
  }

  @Test
  fun test1(){
    assertEquals(setOf("eat", "oath"), findWords(
        arrayOf(
            "oaan",
            "etae",
            "ihkr",
            "iflv"
        ).map{it.toCharArray()}.toTypedArray(),
        arrayOf("oath", "pea", "eat", "rain")
    ).toSet())
  }

  @Test
  fun test2(){
    assertEquals(listOf("a"), findWords(
        arrayOf(
            "a"
        ).map{it.toCharArray()}.toTypedArray(),
        arrayOf("a")
    ))
  }

  @Test
  fun test3(){
    assertEquals(listOf<String>(), findWords(
        arrayOf(
            "aa"
        ).map{it.toCharArray()}.toTypedArray(),
        arrayOf("aaa")
    ))
  }
}